File: App02_Library/app/src/main/java/com/ini/library/FirstFragment.java
package com.ini.library;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import com.ini.library.adapter.BookAdapter;
import com.ini.library.databinding.FragmentFirstBinding;
import com.ini.library.model.BookDTO;
import com.ini.library.service.NaverAPIServiceV1;

import java.util.List;

public class FirstFragment extends Fragment {

    // 'fragment_first.xml을 Binding하라' 라는 의미가 된다.
    // *.xml 파일이 마치 java class가 된 것처럼 코딩을 할 수 있다.
    private FragmentFirstBinding binding;

    private BookAdapter bookAdapter;
    private List<BookDTO> bookDTOList;

    @Override
    public View onCreateView(
            LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState
    ) {

        binding = FragmentFirstBinding.inflate(inflater, container, false);
        return binding.getRoot();

    }

    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        bookAdapter = null; // = new BookAdapter();
        /*
         * Adapter를 생성하고 Adapter와 RecyclerView를 연결하는 코드를 Activity, Fragment와 같은 부분에서 처리를 해야하는 데...
         *
         * NaverAPIServiceV1에게 역할을 대신 수행하도록 한다.
         * 그러기 위해서 Adapter, RecyclerView 등이 담긴 binding 객체를 생성자의 매개변수로 전달한다.
         */
        NaverAPIServiceV1 nService = new NaverAPIServiceV1(bookAdapter, binding);
        nService.getNaverBooks("백신 예방");

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

}

File: HelloApp3/app/src/main/java/com/ini/hello/FirstFragment.kt
package com.ini.hello

import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.ini.hello.databinding.FragmentFirstBinding

class FirstFragment : Fragment() {

    /*
    * _firstFragment가 외부에서 어떤 이유로 값이 변경이 되면 firstFragment 자체에도 문제가 생길 수 있다.
    * 때문에 _firstFragment를 불변 객체로 선언하고 firstFragment만 외부로 보내서 사용을 하게 만든다.
    */

    /**
     * Fragment Binding하기
     * 1. null값으로 초기화하기
     */
    private var _firstFragment: FragmentFirstBinding? = null

    // _firstFragment에 있는 모든 속성을 firstFragment에 복제하는 것
    /**
     * _firstFragment와 fragment의 모든 속성을 연결하여
     * _firstFragment 현재 클래스에서 safe 데이터 형식으로 값을 설정하고
     * 외부로 전송할 때는 get()가 설정된 firstFragment를 사용하는 다소 독특한 방식.
     *
     * 내부 클래스에서 변경되는 변수가 외부로 전달될 때 문제를 일으킬 수 있기 때문에
     * 외부에서는 변수값을 변경하지 못하도록 immutable(불변) 객체로 변환하는 작업
     */
    private val firstFragment get() = _firstFragment!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

//        firstFragment = FragmentFirstBinding.inflate(layoutInflater)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment

        // firstFragment = FragmentFirstBinding.inflate(inflater, container, false)
        _firstFragment = FragmentFirstBinding.inflate(inflater, container, false)

//        return inflater.inflate(R.layout.fragment_first, container, false)
        return firstFragment.root
    }
}
